compiler/rustc_codegen_ssa/src/back/linker.rs

```diff
@@ -276,18 +276,27 @@ impl<'a> Linker for GccLinker<'a> {
     fn set_output_kind(&mut self, output_kind: LinkOutputKind, out_filename: &Path) {
         match output_kind {
             LinkOutputKind::DynamicNoPicExe => {
-                if !self.is_ld && self.sess.target.linker_is_gnu {
+                if !self.is_ld
+                    && self.sess.target.linker_is_gnu
+                    && !self.sess.target.is_like_windows
+                {
                     self.cmd.arg("-no-pie");
                 }
             }
             LinkOutputKind::DynamicPicExe => {
-                // `-pie` works for both gcc wrapper and ld.
-                self.cmd.arg("-pie");
+                // noop on windows w/ gcc & ld, error w/ lld
+                if !self.sess.target.is_like_windows {
+                    // `-pie` works for both gcc wrapper and ld
+                    self.cmd.arg("-pie");
+                }
             }
             LinkOutputKind::StaticNoPicExe => {
                 // `-static` works for both gcc wrapper and ld.
                 self.cmd.arg("-static");
-                if !self.is_ld && self.sess.target.linker_is_gnu {
+                if !self.is_ld
+                    && self.sess.target.linker_is_gnu
+                    && !self.sess.target.is_like_windows
```

Comment:
The `-no-pie` changes here and above are not necessary, it's purely a compiler's (gcc/clang) option (see the the `!self.is_ld` condition), it never reaches the underlying linker.

(It's again one of the cases of `linker_is_gnu` being vague, here its meaning is rather `compiler_is_gnu`.)